src/sqlfluff/core/linter/linted_file.py

```diff
@@ -314,6 +321,21 @@ def fix_string(self) -> Tuple[Any, bool]:
             )
             local_type_list = [slc.slice_type for slc in local_raw_slices]
 
+            if set(local_raw_slices).intersection(raw_slices_in_template_loop):
+                # The patch is being applied to a loop body. Sanity check it.
+                patch_split_list = patch.fixed_raw.split()
+                patch_split_set = set(patch.fixed_raw.split())
+                if len(patch_split_list) - len(patch_split_set) >= 2:
+                    # This is a heuristic. If the patch contains 2 or more
+                    # duplicate strings, assume it's a poorly computed patch
+                    # where the fixes to expanded code couldn't be applied
+                    # cleanly to the source code.
+                    linter_logger.info(
+                        "      - Skipping apparent bad patch inside template loop: %s",
+                        patch,
+                    )
+                    continue
```

Comment:
Seems this could be prone to false positives. Wonder what we can do to reduce that? Could we add lots of test cases?

Reply:
Difficult to say. I think this will only trigger for loops that don't do any templating inside the loop, which, now that I think about it, is a **pretty strange thing to do**. I asked the author of the original issue what he's actually trying to do, because none of his examples generate useful looking code, e.g.:
```
SELECT 2, 2, 2, 2
```

Umm, what?

Ultimately, I think it's hard to solve this 100% robustly without getting deeper into Jinja. As Alan mentioned on the podcast, Jinja doesn't seem to support this. I have a _suspicion_ that if we look really closely at Jinja, we may be able to do some things to make this work, although they may be hacky at first. But if we find something promising, maybe we can submit an upstream PR.

Basically, if we could determine which sections of code came from different iterations of a loop, we could avoid these issues much more reliably and straightforwardly.

Reply:
Some other random thoughts:
* Currently, SQLFluff sometimes corrupts the source code. That's a very severe bug. A code fixer should never _break_ the code.
* In the short run, I'm willing to have SQLFluff make some mistakes of omission (not changing code) or even poor formatting in order to avoid corrupting the code.
* In the long run, I expect we'll see user reports where this isn't quite behaving correctly, and we'll address those.

In a perfect world, I'd take a few engineers and a QA person and we'd go off and work on this for a few months in isolation. In practice, we have limited engineering time, and our users provide a lot of the QA. It's not perfect, but SQLFluff _does_ say it's in beta. I was wondering if we may want our next release to be a [pre-release](https://www.python.org/dev/peps/pep-0440/#pre-releases). That will prevent people from accidentally picking up unstable code, while still giving us a chance to get preliminary feedback from the community.

Reply:
Yeah I'm still veering towards the "if Jinja doesn't support this easily, then maybe we should gracefully bail out as say we can't fix for loop" argument I suggested on the other issue, rather than risk breaking things. I didn't 100% follow this code, but is that what this PR is basically attempting to do by checking this? In which cse maybe bailing out (even if it's more often than it used to) is the "right" thing to do.

I just never like the word "heuristics!" - to me it just suggests magic AI 😁

Reply:
> I was wondering if we may want our next release to be a [pre-release](https://www.python.org/dev/peps/pep-0440/#pre-releases). That will prevent people from accidentally picking up unstable code, while still giving us a chance to get preliminary feedback from the community.

Do we get much feedback for pre-releases? If not, then question the value of doing this. I think it makes sense for certain things (e.g. Alan is proposing a pre-release for moving dbt template to a plugin and can have some the maintainers test that), but for others it seems to just put off the release with the hope of gaining more confidence without any real knowledge that it will give us that confidence...

Reply:
I had the same thought yesterday afternoon (don't do it!), but we have existing tests that rely on this, and if you look at those tests, they are pretty useful things to support. We have code like this at my job -- picture a "for" loop where each iteration expands to a 50-line SELECT that uses just a little templating -- a column name here, a FROM table there. So I see both sides of this -- the user side (I want!) and the engineer side (it's flaky!).

Basically, this code is looking for what appear to be duplicate strings in the expanded code, and if it sees them, discard the patches in that area.

I used the word heuristic somewhat deliberately. To me, it's a fancy word for "guess'. I leave the magic AI to the data scientists on the team I work with. 🤣

Reply:
Resolving as you went a different approach in the end.